src/server.c

```diff
@@ -1667,22 +1667,29 @@ int clientsCronResizeQueryBuffer(client *c) {
     size_t querybuf_size = sdsalloc(c->querybuf);
     time_t idletime = server.unixtime - c->lastinteraction;
 
-    /* There are two conditions to resize the query buffer:
-     * 1) Query buffer is > PROTO_RESIZE_THRESHOLD and too big for latest peak.
-     * 2) Query buffer is > PROTO_RESIZE_THRESHOLD and client is idle. */
-    if (querybuf_size > PROTO_RESIZE_THRESHOLD &&
-        ((c->querybuf_peak && (querybuf_size/(c->querybuf_peak+1)) > 2) ||
-         idletime > 2))
-    {
-        /* Only resize the query buffer if it is actually wasting
-         * at least a few kbytes. */
-        if (sdsavail(c->querybuf) > 1024*4) {
+    /* Only resize the query buffer if the buffer is bigger than
+     * PROTO_RESIZE_THRESHOLD, and it is actually wasting at least a few kbytes. */
+    if (querybuf_size > PROTO_RESIZE_THRESHOLD && sdsavail(c->querybuf) > 1024*4) {
+        /* There are two conditions to resize the query buffer: */
+        if (idletime > 2) {
+            /* 1) Query is idle for a long time. */
             c->querybuf = sdsRemoveFreeSpace(c->querybuf);
+        } else if ((querybuf_size/(c->querybuf_peak+1)) > 2) {
+            /* 2) Query buffer is > PROTO_RESIZE_THRESHOLD and too big for latest peak. */
+             /* TODO: instead of the condition below, we can maybe do trim excess space but only up to a limit.
+              * i.e: c->querybuf = sdsRemoveFreeSpaceTo(c->querybuf, MAX(sdslen(c->querybuf),c->bulklen)); */
+            if (c->bulklen == -1)
+                c->querybuf = sdsRemoveFreeSpace(c->querybuf);
         }
     }
+
     /* Reset the peak again to capture the peak memory usage in the next
      * cycle. */
-    c->querybuf_peak = 0;
+    c->querybuf_peak = sdslen(c->querybuf);
+    /* We reset to either the current used, or currently processed bulk size,
+     * which ever is bigger. */
+    if (c->bulklen != -1 && c->bulklen > (long)c->querybuf_peak)
```

Comment:
Won't it be better to upcast `bulklen` instead of downcasting `querybuf_peak`?
```suggestion
    if (c->bulklen != -1 && (size_t)c->bulklen > c->querybuf_peak)
```

Reply:
they're both the same size in both 32 and 64 bits.. only different is the sign.